early-attack-detector-py/src/agent.py

```diff
@@ -0,0 +1,401 @@
+                    contract_address = calc_contract_address(w3, trace.action.from_, nonce - 1)
+
+                logger.warn(
+                    f"Contract {contract_address} creation failed with tx {trace.transaction_hash}: {error}"
+                )
+            if created_contract_address.lower() in previous_contracts:
+                if ENV == 'dev':
+                    logger.info(f"Contract {created_contract_address} was already created")
+                continue
+            previous_contracts.append(created_contract_address.lower())
+            # creation bytecode contains both initialization and run-time bytecode.
+            creation_bytecode = trace.action.init
+            if creation_bytecode in all_creation_bytecodes.values():
+                if ENV == 'dev':
+                    logger.info(f"Contract {created_contract_address} was already created")
+                original_contract_address = list(all_creation_bytecodes.keys())[list(all_creation_bytecodes.values()).index(creation_bytecode)]
+                if original_contract_address not in repeated_bytecodes.keys():
+                    repeated_bytecodes[original_contract_address] = {}
+                if trace.action.from_ not in repeated_bytecodes[original_contract_address].keys():
+                    repeated_bytecodes[original_contract_address][trace.action.from_] = []
+                repeated_bytecodes[original_contract_address][trace.action.from_].append(created_contract_address)
+                continue
+            all_creation_bytecodes[created_contract_address] = creation_bytecode
+            for finding in detect_malicious_contract(
+                w3,
+                trace.action.from_,
+                created_contract_address,
+                creation_bytecode,
+                error=error,
+            ):
+                finding.addresses = [trace.action.from_, created_contract_address]
+                if finding.severity == FindingSeverity.Critical:
+                    # We priorize when there are critical findings
+                    funding_alerts, funding_labels = check_funding_labels(trace.action.from_, tx_timestamp=tx_timestamp, n_days=FUNDING_TIME,
+                                                                          extra_time_bots=EXTRA_TIME_BOTS, extra_time=EXTRA_TIME_DAYS)
+                    if len(funding_labels) > 0:
+                        finding.metadata["funding_alerts"] = ','.join(funding_alerts)
+                        finding.metadata["funding_labels"] = ','.join(funding_labels)
+                    if float(finding.metadata['model_score']) >= MODEL_PRECISION_THRESHOLD:
+                        finding.metadata['high_precision_model'] = True
+                    # If the model is working in high precision, or it has a 1-day funding alert, we raise the alert and continue
+                    if 'high_precision_model' in finding.metadata.keys() or 'funding_labels' in finding.metadata.keys():
```

Comment:
For high-precision alerts (if they are frequent enough) we could perhaps bypass querying funding labels

Reply:
I think having high precision + funding may also be interesting for easier understanding of the alert